File: Python3/Maximum Product Subarray.py
#1
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        r = nums[0]
        imax = r
        imin = r

        for i in range(1, len(nums)):
            if nums[i] < 0:
                imax, imin = imin, imax
            
            imax = max(nums[i], imax * nums[i])
            imin = min(nums[i], imin * nums[i])

            r = max(r, imax)
        return r
        

#2
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        B = nums[::-1]
        for i in range(1, len(nums)):
            nums[i] *= nums[i - 1] or 1
            B[i] *= B[i - 1] or 1
        return max(nums + B)
        

File: Golang/Maximum Product Subarray.go
func maxProduct(nums []int) int {
    r := nums[0]
    max:=r
    min:=r
    for i:=1;i < len(nums);i++ {
        if nums[i] < 0 {
            max, min = min, max
        }

        max = maxFunc(nums[i], max*nums[i])
        min = minFunc(nums[i], min*nums[i])

        r = maxFunc(r, max)
    }
    return r
}

func maxFunc(a, b int) int {
    if a > b {
        return a
    }
    return b
}

func minFunc(a, b int) int {
    if a > b {
        return b
    }
    return a
}